instrumentation/opentelemetry-instrumentation-openai/src/opentelemetry/instrumentation/openai/patch.py

```diff
@@ -0,0 +1,438 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+from opentelemetry import trace
+from opentelemetry.trace import SpanKind, Span
+from opentelemetry.trace.status import Status, StatusCode
+from opentelemetry.trace.propagation import set_span_in_context
+from openai._types import NOT_GIVEN
+from span_attributes import SpanAttributes, LLMSpanAttributes, Event
+from utils import estimate_tokens, silently_fail, extract_content, calculate_prompt_tokens
+
+
+def chat_completions_create(original_method, version, tracer):
+    """Wrap the `create` method of the `ChatCompletion` class to trace it."""
+
+    def traced_method(wrapped, instance, args, kwargs):
+        llm_prompts = []
+        for item in kwargs.get("messages", []):
+            tools = get_tool_calls(item)
+            if tools is not None:
+                tool_calls = []
+                for tool_call in tools:
+                    tool_call_dict = {
+                        "id": tool_call.id if hasattr(tool_call, "id") else "",
+                        "type": tool_call.type if hasattr(tool_call, "type") else "",
+                    }
+                    if hasattr(tool_call, "function"):
+                        tool_call_dict["function"] = {
+                            "name": (
+                                tool_call.function.name
+                                if hasattr(tool_call.function, "name")
+                                else ""
+                            ),
+                            "arguments": (
+                                tool_call.function.arguments
+                                if hasattr(tool_call.function, "arguments")
+                                else ""
+                            ),
+                        }
+                    tool_calls.append(tool_call_dict)
+                llm_prompts.append(tool_calls)
+            else:
+                llm_prompts.append(item)
+
+        span_attributes = {
+            **get_llm_request_attributes(kwargs, prompts=llm_prompts),
+        }
+
+        attributes = LLMSpanAttributes(**span_attributes)
+
+        span = tracer.start_span(
+            "openai.completion",
+            kind=SpanKind.CLIENT,
+            context=set_span_in_context(trace.get_current_span()),
+        )
+        _set_input_attributes(span, kwargs, attributes)
+
+        try:
+            result = wrapped(*args, **kwargs)
+            if is_streaming(kwargs):
+                prompt_tokens = 0
```

Comment:
Do we need to calculate?

OpenAI now supports returning usage in the last chunk - https://community.openai.com/t/usage-stats-now-available-when-using-streaming-with-the-chat-completions-api-or-completions-api/738156

I'd suggest to remove manual calculation and ask users to enable it on the request options if they need this information.

We can always come back and add it later if it's proven to be necessary.

Reply:
agree, will fix